internal/tofu/eval_context_builtin.go

```diff
@@ -128,14 +128,18 @@ func (ctx *BuiltinEvalContext) Input() UIInput {
 	return ctx.InputValue
 }
 
-func (ctx *BuiltinEvalContext) InitProvider(addr addrs.AbsProviderConfig) (providers.Interface, error) {
+func (ctx *BuiltinEvalContext) InitProvider(addr addrs.AbsProviderConfig, providerKey addrs.InstanceKey) (providers.Interface, error) {
 	ctx.ProviderLock.Lock()
 	defer ctx.ProviderLock.Unlock()
 
 	key := addr.String()
```

Comment:
nit: key naming here is pretty confusing and I think this should be fixed via a greater change involving provider address refactoring (as outlined in the fully-dynamic providers RFC). I am happy to postpone the refactoring, however, it would be nice to rename `providerKey` to `providerInstanceKey` and `key` to `providerConfigName` or something similar to avoid the following:
```
ctx.ProviderCache[key][providerKey] = p
```

Reply:
I think we should create a separate issue to refactor provider addresses as described in the fully dynamic provider iteration RFC regardless of whether or not we are going to implement the whole RFC.

Reply:
@apparentlymart will be updating the tracking issue for this item.